File: bridge/src/vhdr_loader.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use crate::Recording;

enum Bm { I16, U16, I32, F32 }

pub fn load_vhdr(path: &Path) -> Result<Recording, String> {
    let content = fs::read_to_string(path)
        .map_err(|e| format!("Failed reading .vhdr header {}: {}", path.display(), e))?;

    let mut data_file = String::new();
    let mut orientation = String::from("MULTIPLEXED");
    let mut binary_format = String::from("IEEE_FLOAT_32");
    let mut num_channels = 0usize;
    let mut sampling_interval_us = 0.0f64;

    let mut channel_names = HashMap::new();
    let mut channel_resolutions = HashMap::new();

    let mut section = String::new();

    for raw_line in content.lines() {
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with(';') {
            continue;
        }
        if line.starts_with('[') && line.ends_with(']') {
            section = line[1..line.len() - 1].trim().to_string();
            continue;
        }
        if let Some(idx) = line.find('=') {
            let k = line[..idx].trim();
            let v = line[idx + 1..].trim();
            match section.as_str() {
                "Common Infos" => {
                    if k.eq_ignore_ascii_case("DataFile") {
                        data_file = v.to_string();
                    } else if k.eq_ignore_ascii_case("DataOrientation") {
                        orientation = v.to_uppercase();
                    } else if k.eq_ignore_ascii_case("NumberOfChannels") {
                        num_channels = v.parse().unwrap_or(0);
                    } else if k.eq_ignore_ascii_case("SamplingInterval") {
                        sampling_interval_us = v.parse().unwrap_or(0.0);
                    }
                }
                "Binary Infos" => {
                    if k.eq_ignore_ascii_case("BinaryFormat") {
                        binary_format = v.to_uppercase();
                    }
                }
                "Channel Infos" => {
                    if k.to_lowercase().starts_with("ch") {
                        if let Ok(ch_idx) = k[2..].parse::<usize>() {
                            let parts: Vec<&str> = v.split(',').collect();
                            let name = if !parts.is_empty() && !parts[0].trim().is_empty() {
                                parts[0].trim().to_string()
                            } else {
                                format!("Ch{}", ch_idx)
                            };
                            channel_names.insert(ch_idx, name);
                            let res = if parts.len() >= 3 && !parts[2].trim().is_empty() {
                                parts[2].trim().parse::<f32>().unwrap_or(1.0)
                            } else {
                                1.0
                            };
                            channel_resolutions.insert(ch_idx, res);
                        }
                    }
                }
                _ => {}
            }
        }
    }

    if sampling_interval_us <= 0.0 {
        return Err(format!("Invalid SamplingInterval {} in .vhdr", sampling_interval_us));
    }
    let rate = 1_000_000.0 / sampling_interval_us;
    if num_channels == 0 {
        return Err("NumberOfChannels is 0 in .vhdr".to_string());
    }

    let mut labels = Vec::with_capacity(num_channels);
    let mut resolutions = Vec::with_capacity(num_channels);
    for i in 1..=num_channels {
        labels.push(channel_names.get(&i).cloned().unwrap_or_else(|| format!("Ch{}", i)));
        resolutions.push(channel_resolutions.get(&i).cloned().unwrap_or(1.0));
    }

    let parent = path.parent().unwrap_or_else(|| Path::new(""));
    let mut data_path = parent.join(&data_file);
    if data_file.is_empty() || !data_path.exists() {
        let eeg_path = path.with_extension("eeg");
        let dat_path = path.with_extension("dat");
        if eeg_path.exists() {
            data_path = eeg_path;
        } else if dat_path.exists() {
            data_path = dat_path;
        } else {
            return Err(format!("Companion data file not found for {}", path.display()));
        }
    }

    let bytes = fs::read(&data_path)
        .map_err(|e| format!("Failed reading EEG data file {}: {}", data_path.display(), e))?;

    let bytes_per_sample = match binary_format.as_str() {
        "INT_16" | "UINT_16" => 2,
        "INT_32" => 4,
        _ => 4, // IEEE_FLOAT_32
    };

    let bm = match binary_format.as_str() {
        "INT_16" => Bm::I16,
        "UINT_16" => Bm::U16,
        "INT_32" => Bm::I32,
        _ => Bm::F32,
    };

    let total_samples = bytes.len() / (num_channels * bytes_per_sample);
    if total_samples == 0 {
        return Err(format!("Data file {} contains 0 complete samples", data_path.display()));
    }

    let mut channels = vec![vec![0.0f32; total_samples]; num_channels];

    if orientation == "VECTORIZED" {
        for c in 0..num_channels {
            let res = resolutions[c];
            let ch_slice = &mut channels[c];
            for s in 0..total_samples {
                let offset = (c * total_samples + s) * bytes_per_sample;
                if offset + bytes_per_sample <= bytes.len() {
                    let v = match bm {
                        Bm::I16 => i16::from_le_bytes([bytes[offset], bytes[offset + 1]]) as f32,
                        Bm::U16 => u16::from_le_bytes([bytes[offset], bytes[offset + 1]]) as f32,
                        Bm::I32 => i32::from_le_bytes([
                            bytes[offset],
                            bytes[offset + 1],
                            bytes[offset + 2],
                            bytes[offset + 3],
                        ]) as f32,
                        Bm::F32 => f32::from_le_bytes([
                            bytes[offset],
                            bytes[offset + 1],
                            bytes[offset + 2],
                            bytes[offset + 3],
                        ]),
                    };
                    ch_slice[s] = v * res;
                }
            }
        }
    } else {
        // MULTIPLEXED
        let mut offset = 0;
        for s in 0..total_samples {
            for c in 0..num_channels {
                if offset + bytes_per_sample <= bytes.len() {
                    let v = match bm {
                        Bm::I16 => i16::from_le_bytes([bytes[offset], bytes[offset + 1]]) as f32,
                        Bm::U16 => u16::from_le_bytes([bytes[offset], bytes[offset + 1]]) as f32,
                        Bm::I32 => i32::from_le_bytes([
                            bytes[offset],
                            bytes[offset + 1],
                            bytes[offset + 2],
                            bytes[offset + 3],
                        ]) as f32,
                        Bm::F32 => f32::from_le_bytes([
                            bytes[offset],
                            bytes[offset + 1],
                            bytes[offset + 2],
                            bytes[offset + 3],
                        ]),
                    };
                    channels[c][s] = v * resolutions[c];
                    offset += bytes_per_sample;
                }
            }
        }
    }

    Ok(Recording {
        rate,
        labels,
        channels,
        source_epoch_samples: None,
        epoch_labels: None,
    })
}

File: lib/src/vhdr_loader.dart
import 'dart:io';
import 'dart:typed_data';

import 'models.dart';

/// Loader for BrainVision BrainAmp format files (.vhdr / .eeg / .dat).
///
/// Supports:
///   DataOrientation : MULTIPLEXED | VECTORIZED
///   BinaryFormat    : IEEE_FLOAT_32 | INT_16 | UINT_16 | INT_32
///   Resolution      : per-channel float multiplier from [Channel Infos]
class VhdrLoader {
  EegRecording load(String vhdrPath) {
    // ── Parse header ───────────────────────────────────────────────────────
    final lines = File(vhdrPath)
        .readAsStringSync()
        .replaceAll('\r\n', '\n')
        .replaceAll('\r', '\n')
        .split('\n');

    String key(String line) =>
        line.contains('=') ? line.substring(0, line.indexOf('=')).trim() : '';
    String val(String line) =>
        line.contains('=') ? line.substring(line.indexOf('=') + 1).trim() : '';

    String dataFile = '';
    String orientation = 'MULTIPLEXED';
    String binaryFormat = 'IEEE_FLOAT_32';
    int numChannels = 0;
    int samplingIntervalUs = 0; // microseconds

    final channelNames = <int, String>{};
    final channelResolutions = <int, double>{}; // index → µV/bit

    String section = '';
    for (final raw in lines) {
      final line = raw.trim();
      if (line.isEmpty || line.startsWith(';')) continue;
      if (line.startsWith('[')) {
        section = line.replaceAll('[', '').replaceAll(']', '').trim();
        continue;
      }
      final k = key(line);
      final v = val(line);
      switch (section) {
        case 'Common Infos':
          if (k == 'DataFile') dataFile = v;
          if (k == 'DataOrientation') orientation = v.toUpperCase();
          if (k == 'NumberOfChannels') numChannels = int.tryParse(v) ?? 0;
          if (k == 'SamplingInterval') samplingIntervalUs = int.tryParse(v) ?? 0;
        case 'Binary Infos':
          if (k == 'BinaryFormat') binaryFormat = v.toUpperCase();
        case 'Channel Infos':
          // Ch<N>=name,ref,resolution,unit,...
          if (k.startsWith('Ch')) {
            final idx = int.tryParse(k.substring(2));
            if (idx != null) {
              final parts = v.split(',');
              channelNames[idx] = parts.isNotEmpty ? parts[0].trim() : 'Ch$idx';
              if (parts.length >= 3 && parts[2].trim().isNotEmpty) {
                channelResolutions[idx] = double.tryParse(parts[2].trim()) ?? 1.0;
              } else {
                channelResolutions[idx] = 1.0;
              }
            }
          }
      }
    }

    if (samplingIntervalUs <= 0) {
      throw const FormatException('Invalid SamplingInterval in .vhdr');
    }
    final sampleRate = 1000000.0 / samplingIntervalUs;
    if (numChannels <= 0) {
      throw FormatException('Invalid NumberOfChannels=$numChannels in .vhdr');
    }

    // Build ordered label list
    final labels = <String>[];
    final resolutions = <double>[];
    for (var i = 1; i <= numChannels; i++) {
      labels.add(channelNames[i] ?? 'Ch$i');
      resolutions.add(channelResolutions[i] ?? 1.0);
    }

    // ── Resolve data file path ──────────────────────────────────────────────
    final vhdrDir = File(vhdrPath).parent.path;
    final dataPath = dataFile.isNotEmpty
        ? '$vhdrDir${Platform.pathSeparator}${dataFile.split('/').last}'
        : vhdrPath.replaceAll(RegExp(r'\.vhdr$', caseSensitive: false), '.eeg');
    if (!File(dataPath).existsSync()) {
      throw FileSystemException('Data file not found: $dataPath', dataPath);
    }

    // ── Read binary data ────────────────────────────────────────────────────
    final bytes = File(dataPath).readAsBytesSync();
    final bytesPerSample = switch (binaryFormat) {
      'INT_16' || 'UINT_16' => 2,
      'INT_32' => 4,
      _ => 4, // IEEE_FLOAT_32
    };
    final totalSamples = bytes.length ~/ (numChannels * bytesPerSample);

    if (totalSamples <= 0) {
      throw FormatException('Data file appears empty: $dataPath');
    }

    final bd = ByteData.sublistView(bytes);
    const stride = 1;

    final preview = <Float32List>[];

    if (orientation == 'VECTORIZED') {
      // Each channel is stored contiguously: [ch0_t0, ch0_t1, ..., ch1_t0, ...]
      for (var c = 0; c < numChannels; c++) {
        final res = resolutions[c];
        final channelValues = <double>[];
        for (var s = 0; s < totalSamples; s += stride) {
          final offset = (c * totalSamples + s) * bytesPerSample;
          channelValues.add(_readSample(bd, offset, binaryFormat, res));
        }
        preview.add(Float32List.fromList(channelValues));
      }
    } else {
      // MULTIPLEXED: [ch0_t0, ch1_t0, ..., chN_t0, ch0_t1, ch1_t1, ...]
      for (var c = 0; c < numChannels; c++) {
        final res = resolutions[c];
        final channelValues = <double>[];
        for (var s = 0; s < totalSamples; s += stride) {
          final offset = (s * numChannels + c) * bytesPerSample;
          channelValues.add(_readSample(bd, offset, binaryFormat, res));
        }
        preview.add(Float32List.fromList(channelValues));
      }
    }

    final markers = _loadVmrk(vhdrPath, sampleRate);

    return EegRecording(
      path: vhdrPath,
      dataPath: dataPath,
      sampleRate: sampleRate,
      labels: labels,
      preview: preview,
      sampleCount: totalSamples,
      format: 'vhdr',
      markers: markers,
    );
  }

  List<EegMarker> _loadVmrk(String vhdrPath, double sampleRate) {
    final vmrkPath =
        vhdrPath.replaceAll(RegExp(r'\.vhdr$', caseSensitive: false), '.vmrk');
    final file = File(vmrkPath);
    if (!file.existsSync()) return const [];

    final markers = <EegMarker>[];
    final lines = file
        .readAsStringSync()
        .replaceAll('\r\n', '\n')
        .replaceAll('\r', '\n')
        .split('\n');

    String section = '';
    for (final raw in lines) {
      final line = raw.trim();
      if (line.isEmpty || line.startsWith(';')) continue;
      if (line.startsWith('[')) {
        section = line.replaceAll('[', '').replaceAll(']', '').trim();
        continue;
      }
      if (section == 'Marker Infos' && line.startsWith('Mk')) {
        final eqIdx = line.indexOf('=');
        if (eqIdx < 0) continue;
        final val = line.substring(eqIdx + 1).trim();
        final parts = val.split(',');
        if (parts.length >= 3) {
          final type = parts[0].trim();
          final desc = parts[1].trim();
          final posSamples = int.tryParse(parts[2].trim()) ?? 1;
          final sizeSamples =
              parts.length >= 4 ? (int.tryParse(parts[3].trim()) ?? 0) : 0;
          final chNum =
              parts.length >= 5 ? (int.tryParse(parts[4].trim()) ?? 0) : 0;

          final startSec = (posSamples - 1) / sampleRate;
          final durSec = sizeSamples / sampleRate;

          markers.add(EegMarker(
            type: type,
            description: desc,
            startSeconds: startSec > 0 ? startSec : 0.0,
            durationSeconds: durSec > 0 ? durSec : 0.0,
            channelIndex: chNum > 0 ? chNum - 1 : null,
          ));
        }
      }
    }
    return markers;
  }


  double _readSample(ByteData bd, int offset, String format, double resolution) {
    if (offset + (format == 'INT_16' || format == 'UINT_16' ? 2 : 4) > bd.lengthInBytes) {
      return 0.0;
    }
    return switch (format) {
      'IEEE_FLOAT_32' => bd.getFloat32(offset, Endian.little),
      'INT_16' => bd.getInt16(offset, Endian.little) * resolution,
      'UINT_16' => bd.getUint16(offset, Endian.little) * resolution,
      'INT_32' => bd.getInt32(offset, Endian.little) * resolution,
      _ => bd.getFloat32(offset, Endian.little),
    };
  }
}
